pending.c

```diff
@@ -169,21 +339,28 @@ void w_pending_coll_append(struct watchman_pending_collection *target,
   struct watchman_pending_fs *p, *target_p;
 
   while ((p = w_pending_coll_pop(src)) != NULL) {
-    target_p = w_ht_val_ptr(w_ht_get(target->pending_uniq,
-                            w_ht_ptr_val(p->path)));
+    target_p = art_search(&target->tree,
+        (const uint8_t *)p->path->buf, p->path->len);
     if (target_p) {
       /* Entry already exists: consolidate */
-      consolidate_item(target_p, p->flags);
+      consolidate_item(target, target_p, p->flags);
       w_pending_fs_free(p);
       continue;
     }
 
-    p->next = target->pending;
-    target->pending = p;
-    w_ht_set(target->pending_uniq, w_ht_ptr_val(p->path), w_ht_ptr_val(p));
+    if (is_obsoleted_by_containing_dir(target, p->path)) {
+      w_pending_fs_free(p);
+      continue;
+    }
+    prune_obsoleted_children(target, p->path, p->flags);
```

Comment:
One general comment here is that it is probably possible to split this up so that `is_obsoleted_by_containing_dir` is in a separate diff from the `prune_obsoleted_children` function (both really seem to add on to each other). But it's ok :)